File: Queue/C/PriQueue.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    int prio;
    struct node * next;
    struct node * prev;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

void enqueue(PPNODE first, PPNODE last, int no, int pri)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->data = no;
    newn->prio = pri;
    newn->next = NULL;
    newn->prev = NULL;

    if((*first == NULL) && (*last == NULL))
    {
        *first = newn;
        *last = newn;
    }
    else
    {
        (*last)->next = newn;
        newn->prev = *last;

        *last = newn;   
    }
}

int dequeue(PPNODE first, PPNODE last)
{
    int value = 0;
    PNODE iMax = NULL;
    PNODE temp = NULL;

    if((*first == NULL) && (*last == NULL))
    { 
        printf("Unable to delete Queue is empty\n");
        return -1;
    }
    else if (*first == *last)
    {
        value = (*first)->data;
        free(*first);

        *first = NULL;
        *last = NULL;
    }
    else
    {
        iMax = *first;
        temp = *first;

        while(temp != NULL)
        {
            if (iMax->prio < temp->prio)
            {
                iMax = temp;
            }

            temp = temp->next;
        }

        value = iMax->data;

        if(iMax == *first) 
        {
            *first = (*first)->next;
            (*first)->prev = NULL;

            free(iMax);
        }
        else if(iMax == *last) 
        {
            *last = (*last)->prev;
            (*last)->next = NULL;

            free(iMax);
        }
        else
        {
            iMax->next->prev = iMax->prev;
            iMax->prev->next = iMax->next;

            free(iMax);
        }
    }

    return value;
}

void Display(PNODE first, PNODE last)
{
    if((first == NULL) && (last == NULL))
    {
        printf("Queue is empty\n");
        return;
    }

    printf("NULL =");

    while(first != NULL)
    {
        printf(" | %d | %d | = ",first->data, first->prio);
        first = first->next;
    }

    printf("NULL\n");
}

int Count(PNODE first, PNODE last)
{
    int iCount = 0;

    if((first == NULL) && (last == NULL))
    {
        printf("Queue is empty\n");
        return 0;
    }

    while(first != NULL)
    {
        iCount++;
        first = first->next;
    }

    return iCount;
}

int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;
    int iRet = 0;
    int choice = 0;
    int value = 0;
    int prio = 0;

    printf("\n=============================================================\n");
    printf("                       Priority Queue Manager                    \n");
    printf("=============================================================\n");

    while(1)
    {
        printf("\n-------------------------------------------------------------\n");
        printf("Please select an operation:\n");
        printf("-------------------------------------------------------------\n");
        printf(" 1 : Enqueue (Enter data with priority)\n");
        printf(" 2 : Dequeue (Remove element with highest priority)\n");
        printf(" 3 : Display queue elements\n");
        printf(" 4 : Count total elements\n");
        printf(" 0 : Exit\n");
        printf("-------------------------------------------------------------\n");

        printf("Enter your choice: \n");
        scanf("%d",&choice);

        if(choice == 1)
        {
            printf("Enter element to enqueue : \n");
            scanf("%d",&value);
            printf("Enter priority of the element: \n");
            scanf("%d",&prio);
            enqueue(&head, &tail, value, prio);
        }
        else if(choice == 2)
        {
            value = dequeue(&head, &tail);          // "If multiple jobs have same highest priority, the first inserted is removed first."

            if(value != -1)
            {
                printf("Dequeued element with highest priority: %d\n",value);
            }
        }
        else if(choice == 3)
        {
            Display(head, tail);
        }
        else if(choice == 4)
        {
            iRet = Count(head, tail);
            printf("Total elements in deque: %d\n",iRet);
        }
        else if(choice == 0)
        {
            printf("Thank you for using our Application!\n");
            break;
        }
        else
        {
            printf("Invalid choice. Please try again.\n");
        }
    }

    return 0;
};

File: Queue/Python/PriQueue.py

class node:
    def __init__(self, no, pri):
        self.data = no
        self.prio = pri
        self.next = None
        self.prev = None

class PriQueue:

    def __init__(self):
        self.first = None
        self.last = None
        self.iCount = 0

    def enqueue(self, no, pri):
        newn = node(no, pri)

        if(self.first == None and self.last == None):

            self.first = newn
            self.last = newn

        else:

            self.last.next = newn
            newn.prev = self.last

            self.last = newn

        self.iCount += 1

    def dequeue(self):

        if(self.first == None and self.last == None):
            print("Unable to delete Queue is empty")
            return None

        elif(self.first == self.last):
            iValue = self.first.data
            del self.first

            self.first = None
            self.last = None

        else:
            iMax = self.first
            temp = self.first

            while(temp != None):
                if(iMax.prio < temp.prio):
                    iMax = temp

                temp = temp.next

            iValue = iMax.data

            if(iMax == self.first):
                self.first = self.first.next
                self.first.prev = None

            elif(iMax == self.last):
                self.last = self.last.prev
                self.last.next = None

            else:
                iMax.next.prev = iMax.prev
                iMax.prev.next = iMax.next

            del iMax

        self.iCount -= 1
        return iValue


    def Display(self):

        if(self.first == None):
            print("Queue is empty")
            return

        temp = self.first

        print("None <=> ",end="")

        while(temp != None):
            print(f"| {temp.data} | {temp.prio} | <=> ",end="");
            temp = temp.next

        print("None")


    def Count(self):
        return self.iCount


def main():
    obj = PriQueue()

    print("\n=============================================================");
    print("                     Priority Queue Manager                    ");
    print("=============================================================");

    while True:
        print("\n-------------------------------------------------------------")
        print("Please select an operation:")
        print("-------------------------------------------------------------")
        print(" 1 : Enqueue (Enter data with priority)")
        print(" 2 : Dequeue (Remove element with highest priority)")
        print(" 3 : Display queue elements")
        print(" 4 : Count total elements")
        print(" 0 : Exit")
        print("-------------------------------------------------------------")

        choice = int(input("Enter your choice : "))

        if choice == 1:
            value = int(input("Enter element to enqueue: "))
            prio = int(input("Enter priority of the element: "))
            obj.enqueue(value, prio)

        elif choice == 2:
            iRet = obj.dequeue()            # "If multiple jobs have same highest priority, the first inserted is removed first."

            if iRet is not None:
                print(f"Dequeued element with highest priority: {iRet}")

        elif choice == 3:
            obj.Display()

        elif choice == 4:
            print(f"Total elements in Priority Queue: {obj.Count()}")

        elif choice == 0:
            print("Thank you for using our Application!")
            break

        else:
            print("Invalid choice, please try again")


if __name__ == "__main__":
    main()